src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs

```diff
@@ -191,52 +195,33 @@ internal class SqlServerFhirDataStore : IFhirDataStore, IProvideCapability
                         resource.Version = (existingVersion + 1).Value.ToString(CultureInfo.InvariantCulture);
                     }
 
-                    _compressedRawResourceConverter.WriteCompressedRawResource(stream, resource.RawResource.Data);
-
-                    stream.Seek(0, 0);
-
-                    _logger.LogInformation("Upserting {ResourceTypeName} with a stream length of {StreamLength}", resource.ResourceTypeName, stream.Length);
-
-                    // throwing ServiceUnavailableException in order to send a 503 error message to the client
-                    // indicating the server has a transient error, and the client can try again
-                    if (stream.Length < 31) // rather than error on a length of 0, a stream with a small number of bytes should still throw an error. RawResource.Data = null, still results in a stream of 29 bytes
-                    {
-                        _logger.LogCritical("Stream size for resource of type: {ResourceTypeName} is less than 50 bytes, request method: {RequestMethod}", resource.ResourceTypeName, resource.Request.Method);
-                        throw new ServiceUnavailableException();
-                    }
-
-                    PopulateUpsertResourceCommand(sqlCommandWrapper, resource, keepHistory, existingVersion, stream, _coreFeatures.SupportsResourceChangeCapture);
-
                     try
                     {
-                        var newVersion = (int?)await sqlCommandWrapper.ExecuteScalarAsync(cancellationToken);
-                        if (newVersion == null)
-                        {
-                            // indicates a redundant delete
-                            return null;
-                        }
-
-                        if (newVersion == -1)
+                        if (_schemaInformation.Current >= SchemaVersionConstants.Merge)
                         {
-                            // indicates that resource content is same - no new version was created
-                            // We need to send the existing resource in the response matching the correct versionId and lastUpdated as the one stored in DB
-                            return new UpsertOutcome(existingResource, SaveOutcomeType.Updated);
-                        }
-
-                        resource.Version = newVersion.ToString();
-
-                        SaveOutcomeType saveOutcomeType;
-                        if (newVersion == 1)
-                        {
-                            saveOutcomeType = SaveOutcomeType.Created;
+                            var surrId = (await MergeResourcesBeginTransactionAsync(1, cancellationToken)).MinResourceSurrogateId;
```

Comment:
Should we always use 1 as the resourceVersionCount?

Reply:
For current Upsert it is 1. Once we switch to batch it will be number of resource versions in a batch.